mwan3otb/files/usr/sbin/track.lua

```diff
@@ -63,113 +62,76 @@ local function handle_exit()
 	os.exit();
 end
 
-function create_tcp(interface)
-	local tcp = socket.tcp()
-	local fd = p.socket(p.AF_INET, p.SOCK_STREAM, 0)
+function create_socket(interface, kind)
+	local s, fd
+	if kind == "stream" then
+		s = socket.tcp()
+		fd = p.socket(p.AF_INET, p.SOCK_STREAM, 0)
+	elseif kind == "datagram" then
+		s = socket.udp()
+		fd = p.socket(p.AF_INET, p.SOCK_DGRAM, 0)
+	else
+		log("create_socket: unknown kind")
+		return nil
+	end
+	-- TODO: s:bind with ip
 	local ok, err = p.setsockopt(fd, p.SOL_SOCKET, p.SO_BINDTODEVICE, interface)
 	if not ok then
-		log("create_tcp: "..err)
+		log("create_socket: "..err)
 		return nil
 	end
-	tcp:setfd(fd)
-	return tcp
+	s:setfd(fd)
+	return s
 end
 
-function dns_request( host, interface, timeout, domain)
-	local fd, err = p.socket(p.AF_INET, p.SOCK_DGRAM, 0)
-	if not fd then return fd, err end
-
-	p.bind (fd, { family = p.AF_INET, addr = "0.0.0.0", port = 0 })
-
-	-- timeout on socket
-	local ok, err = p.setsockopt(fd, p.SOL_SOCKET, p.SO_RCVTIMEO, math.floor(timeout/1000), (timeout % 1000) * 1000 )
-	if not ok then return ok, err end
-
-	-- bind to specific device
-	local ok, err = p.setsockopt(fd, p.SOL_SOCKET, p.SO_BINDTODEVICE, interface)
-	if not ok then return ok, err end
-
-	local data = dns.encode {
-		id       = math.random(0xFFFF),
-		query    = true,
-		rd       = true,	-- recursion desired
-		opcode   = 'query',
-		question = {
-			name  = domain,	-- FQDN required
-			type  = "A",	-- LOC rr
-			class = "IN"
-		}
-	}
-
+function dns_request(host, interface, timeout)
+	local s = create_socket(interface, "datagram")
+	if not s then
+		return false, "dns_request: no socket"
+	end
+	s:settimeout(timeout)
+	local ok, err = s:setpeername(host, "53")
+	if not ok then
+		s:close()
+		return false, "dns_request: "..err
+	end
+	local id = string.char(math.random(0xFF), math.random(0xFF))
+	local ok, err = s:send(id.."\1 \0\1\0\0\0\0\0\1\tlocalhost\0\0\1\0\1\0\0)\20\0\0\0\0\0\0\0")
```

Comment:
In that case, I would prefer a send("hello this a fake dns request") on a fake dns server.